velox/common/base/PeriodicStatsReportDaemon.cpp

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/PeriodicStatsReportDaemon.h"
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/StatsReporter.h"
+#include "velox/common/caching/AsyncDataCache.h"
+#include "velox/common/caching/CacheTTLController.h"
+#include "velox/common/caching/SsdFile.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/common/memory/MmapAllocator.h"
+
+namespace {
+#define REPORT_IF_NOT_ZERO(name, counter)   \
+  if ((counter) != 0) {                     \
+    RECORD_METRIC_VALUE((name), (counter)); \
+  }
+} // namespace
+
+namespace facebook::velox {
+
+PeriodicStatsReportDaemon::PeriodicStatsReportDaemon(
+    const velox::memory::MemoryAllocator* const memoryAllocator,
+    const velox::cache::AsyncDataCache* const asyncDataCache,
+    const velox::memory::MemoryArbitrator* const arbitrator)
+    : memoryAllocator_(memoryAllocator),
+      asyncDataCache_(asyncDataCache),
+      arbitrator_(arbitrator) {}
+
+void PeriodicStatsReportDaemon::start() {
+  addLowFrequencyReports();
+  addHighFrequencyReports();
+}
+
+void PeriodicStatsReportDaemon::stop() {
+  scheduler_.stop();
+}
+
+void PeriodicStatsReportDaemon::addHighFrequencyReports() {
+  static constexpr uint64_t kHighFrequencyReportIntervalMs{2'000};
+  scheduler_.add("periodic-stats-report.high-frequency", [this]() noexcept {
+    try {
+      reportAllocatorStats();
+    } catch (const std::exception& e) {
+      LOG(ERROR) << "Error reporting high frequency stats : " << e.what();
+    }
+    return std::chrono::milliseconds(kHighFrequencyReportIntervalMs);
+  });
+}
+
+void PeriodicStatsReportDaemon::addLowFrequencyReports() {
+  static constexpr uint64_t kLowFrequencyReportIntervalMs{60'000};
```

Comment:
Could we just let each component decides its own frequency and register separately? Thanks!